README.md
# Unity-on-Gnome
unity style with gnome power

HOW TO INSTALL:

Just copy and paste this commands in terminal

```bash
wget https://raw.githubusercontent.com/RabbitHols/Unity-on-Gnome/master/installer.sh
chmod +x installer.sh 
./installer.sh
```
SCREENSHOTS
![alt text](https://raw.githubusercontent.com/RabbitHols/Unity-on-Gnome/master/screenshots/home.png)
![alt text](https://i.imgur.com/Im5pJ6p.png)
![alt text](https://raw.githubusercontent.com/RabbitHols/Unity-on-Gnome/master/screenshots/dash.png)
![alt text](https://raw.githubusercontent.com/RabbitHols/Unity-on-Gnome/master/screenshots/busy.png)
![alt text](https://i.imgur.com/XJzRy1S.png)

CREDITS
This script uses parts of code from the extensions:<br/>

[Unite-Shell](https://github.com/hardpixel/unite-shell)<br/>
[GnomeGlobalAppMenu](https://extensions.gnome.org/extension/1250/gnome-global-application-menu/)<br/>
[Activities Configurator](https://extensions.gnome.org/extension/358/activities-configurator/)<br/>
[User Theme](https://extensions.gnome.org/extension/19/user-themes/)<br/>

This script uses parts of code from these themes:<br/>

[United-GNOME (GTK Theme)](https://www.gnome-look.org/p/1174889/)<br/>
[Unity8 - RE_DESINGED (Gnome Shell Theme)](https://www.gnome-look.org/p/1206490/)<br/>
[Flat Remix (Icons)](https://www.gnome-look.org/p/1012430/)<br/>
[Capitaine-cursors (Cursor)](https://github.com/keeferrourke/capitaine-cursors)<br/>

To enable extensions from terminal this script use:<br/>

[Gnome-shell-extension-cl](https://github.com/cyberalex4life/gnome-shell-extension-cl)<br/>



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.0.1] - 2018-09-05
### Added
- System update icon in top bar
- App folders 
- Update System icon

### Changed
- Space between icons in top bar
- Removed custom wallpaper

## [0.0.2] - 2018-10-30
- Added compatibility with Java Swing Menu 


debug_installer.sh
#!/bin/sh

sudo add-apt-repository universe
sudo apt update


sudo apt install git gnome-tweak-tool gir1.2-clutter-1.0 gir1.2-clutter-gst-3.0 gir1.2-gtkclutter-1.0 -y
sudo apt install git -y

#Dependency needed for global menu
sudo apt install unity-gtk2-module unity-gtk3-module -y
sudo apt install x11-utils -y

#Needed to recompile gschema of gnome extension with custom default settings
sudo apt-get install libglib2.0-dev -y


git clone https://github.com/RabbitHols/Power-Gnome $HOME/Power-Gnome


sudo cp -r $HOME/Power-Gnome/src/style/themes/* /usr/share/themes
sudo cp -r $HOME/Power-Gnome/src/style/icons/* /usr/share/icons
sudo cp -r $HOME/Power-Gnome/src/style/cursor/* /usr/share/icons

sudo cp -r $HOME/Power-Gnome/src/style/themes/* /usr/share/themes
sudo cp -r $HOME/Power-Gnome/src/style/icons/* /usr/share/icons
sudo cp -r $HOME/Power-Gnome/src/style/cursor/* /usr/share/icons


mkdir $HOME/.local/share/gnome-shell/extensions

sudo yes | cp -rf $HOME/Power-Gnome/src/tweak_extension/user-theme@gnome-shell-extensions.gcampax.github.com $HOME/.local/share/gnome-shell/extensions
sudo yes | cp -rf $HOME/Power-Gnome/src/tweak_extension/activities-config@nls1729 $HOME/.local/share/gnome-shell/extensions

sudo mkdir -p ~/.local/share/gnome-shell/extensions/

#Global Menu
git clone https://gitlab.com/lestcape/Gnome-Global-AppMenu
sudo yes | cp -rf Gnome-Global-AppMenu/gnomeGlobalAppMenu@lestcape ~/.local/share/gnome-shell/extensions/
sudo rm -rf Gnome-Global-AppMenu

sudo rm ~/.local/share/gnome-shell/extensions/gnomeGlobalAppMenu@lestcape/settings-schema.json
sudo cp $HOME/Power-Gnome/src/tweak_extension/gnomeGlobalAppMenu@lestcape/settings-schema.json ~/.local/share/gnome-shell/extensions/gnomeGlobalAppMenu@lestcape/


#Unite
sudo yes | cp -rf $HOME/Power-Gnome/src/tweak_extension/unite@hardpixel.eu ~/.local/share/gnome-shell/extensions/

#Update indicator
sudo yes | cp -rf $HOME/Power-Gnome/src/tweak_extension/apt-update-indicator@franglais125.gmail.com ~/.local/share/gnome-shell/extensions/

#Status area horizontal
sudo yes | cp -rf $HOME/Power-Gnome/src/tweak_extension/status-area-horizontal-spacing@mathematical.coffee.gmail.com ~/.local/share/gnome-shell/extensions/

#App Folder
sudo yes | cp -rf $HOME/Power-Gnome/src/tweak_extension/appfolders-manager@maestroschan.fr ~/.local/share/gnome-shell/extensions/


sudo cp $HOME/Power-Gnome/src/scripts/* /usr/bin
sudo chmod +x /usr/bin/extension-enabler



extension-enabler -e user-theme@gnome-shell-extensions.gcampax.github.com
extension-enabler -e activities-config@nls1729
extension-enabler -e unite@hardpixel.eu
extension-enabler -e gnomeGlobalAppMenu@lestcape
extension-enabler -e apt-update-indicator@franglais125.gmail.com
extension-enabler -e status-area-horizontal-spacing@mathematical.coffee.gmail.com
extension-enabler -e appfolders-manager@maestroschan.fr

sudo cp $HOME/.local/share/gnome-shell/extensions/user-theme@gnome-shell-extensions.gcampax.github.com/schemas/org.gnome.shell.extensions.user-theme.gschema.xml /usr/share/glib-2.0/schemas && sudo glib-compile-schemas /usr/share/glib-2.0/schemas

gsettings set org.gnome.desktop.interface gtk-theme "PowerGnomeGtk" 
gsettings set org.gnome.shell.extensions.user-theme name "PowerGnomeShell"
gsettings set org.gnome.desktop.interface icon-theme 'power-gnome-icons'
gsettings set org.gnome.desktop.interface cursor-theme "power-gnome-cursor"

dconf reset -f /org/gnome/shell/extensions/dash-to-dock/
gsettings set org.gnome.shell.extensions.dash-to-dock show-apps-at-top true
gsettings set org.gnome.shell.extensions.dash-to-dock custom-theme-shrink false

gsettings set org.gnome.desktop.wm.preferences button-layout 'close,maximize,minimize:'
gsettings set org.gnome.shell.extensions.dash-to-dock dash-max-icon-size 30


sudo mv $HOME/Power-Gnome/src/style/wallpaper/powerGnome.png /usr/share/backgrounds/warty-final-ubuntu.png
gsettings set org.gnome.desktop.background picture-uri file:////usr/share/backgrounds/warty-final-ubuntu.png

#Adjust schema error

sudo cp $HOME/.local/share/gnome-shell/extensions/activities-config@nls1729/schemas/org.gnome.shell.extensions.activities-config.gschema.xml /usr/share/glib-2.0/schemas/ 
sudo glib-compile-schemas /usr/share/glib-2.0/schemas/

sudo cp $HOME/.local/share/gnome-shell/extensions/unite@hardpixel.eu/schemas/org.gnome.shell.extensions.unite.gschema.xml /usr/share/glib-2.0/schemas/ 
sudo glib-compile-schemas /usr/share/glib-2.0/schemas/

sudo rm -r Power-Gnome

dbus-send --type=method_call --print-reply --dest=org.gnome.Shell /org/gnome/Shell org.gnome.Shell.Eval string:'global.reexec_self()'
sleep 5s && dbus-send --type=method_call --print-reply --dest=org.gnome.Shell /org/gnome/Shell org.gnome.Shell.Eval string:'global.reexec_self()'
clear
gnome-session-quit
echo Done !!


src/tweak_extension/windowsTab@daniellerch.me/extension.js

/* Based on code from windowsNavigator@gnome-shell-extensions.gcampax.github.com */

const Clutter = imports.gi.Clutter;
const Lang = imports.lang;
const Mainloop = imports.mainloop;
const St = imports.gi.St;

const Main = imports.ui.main;
const Workspace = imports.ui.workspace;
const WorkspacesView = imports.ui.workspacesView;

let current_window_id=1;

function injectToFunction(parent, name, func) {
	let origin = parent[name];
	parent[name] = function() {
		let ret;
		ret = origin.apply(this, arguments);
		if (ret === undefined)
			ret = func.apply(this, arguments);
		return ret;
	}
	return origin;
}

let winInjections, workspaceInjections, workViewInjections, createdActors, connectedSignals;

function resetState() {
	winInjections = { };
	workspaceInjections = { };
	workViewInjections = { };
	createdActors = [ ];
	connectedSignals = [ ];
}

function enable() {
	resetState();

	Workspace.WindowOverlay.prototype.setId = function(id) {
		if (this._text && this._text.visible && id == null)
			this._text.hide();
		this._id = id;
		if (id != null)
			this._text.text = this._id.toString();
	}
	winInjections['setId'] = undefined;


	Workspace.WindowOverlay.prototype.getId = function() {
		return this._id;
	}
	winInjections['getId'] = undefined;

	Workspace.WindowOverlay.prototype.showTooltip = function() {
		if (this._id == null)
			return;
		this._text.raise_top();
		this._text.show();

		if (this._id == current_window_id)
		{
			this._text.text = "Select Window";
		}
		else
		{
			this._text.hide();
		}


	}
	winInjections['showTooltip'] = undefined;


	Workspace.Workspace.prototype.showTooltip = function() {
		if (this._tip == null)
			return;
		this._tip.text = (this.metaWorkspace.index() + 1).toString();
		this._tip.x = this._x;
		this._tip.y = this._y;
		this._tip.show();
		this._tip.raise_top();
	}
	workspaceInjections['showTooltip'] = undefined;


	Workspace.Workspace.prototype.getWindowWithTooltip = function(id) {
		for (let i in this._windowOverlays) {
			if (this._windowOverlays[i] == null)
				continue;
			if (this._windowOverlays[i].getId() === id)
				return this._windowOverlays[i]._windowClone.metaWindow;
		}
		return null;
	}
	workspaceInjections['getWindowWithTooltip'] = undefined;

	Workspace.Workspace.prototype.showWindowsTooltips = function() {

		current_window_id++;
		if(current_window_id>this._windows.length)
			current_window_id=1;

		for (let i in this._windowOverlays) {

			if (this._windowOverlays[i] != null)
			{
				this._windowOverlays[i].showTooltip();
			}
		}
	}
	workspaceInjections['showWindowsTooltips'] = undefined;


	WorkspacesView.WorkspacesView.prototype._onKeyRelease = function(s, o) {

	}
	workViewInjections['_onKeyRelease'] = undefined;

	WorkspacesView.WorkspacesView.prototype._onKeyPress = function(s, o) {
		if(Main.overview._viewSelector._activeTab.id != 'windows')
			return false;

		if (o.get_key_symbol() == Clutter.KEY_Tab && !this._pickWorkspace) {
			this._prevFocusActor = global.stage.get_key_focus();
			global.stage.set_key_focus(null);
			this._active = global.screen.get_active_workspace_index();
			this._pickWindow = true;
			this._workspaces[global.screen.get_active_workspace_index()].showWindowsTooltips();
			return true;
		}


		if (global.stage.get_key_focus() != global.stage)
			return false;

		if (this._pickWindow) {
			let win = this._workspaces[this._active].getWindowWithTooltip(current_window_id);
			if (win)
				Main.activateWindow(win, global.get_current_time());

			return true;
		}


		return false;
	}
	workViewInjections['_onKeyPress'] = undefined;

	winInjections['_init'] = injectToFunction(Workspace.WindowOverlay.prototype, '_init', function(windowClone, parentActor) {
			this._id = null;
			createdActors.push(this._text = new St.Label({ style_class: 'extension-windowsTab-window-tooltip' }));
			this._text.hide();
			parentActor.add_actor(this._text);
			});

	winInjections['updatePositions'] = injectToFunction(Workspace.WindowOverlay.prototype, 'updatePositions', function(cloneX, cloneY, cloneWidth, cloneHeight) {
			let textX = cloneX-50;
			let textY = cloneY-50;
			this._text.set_position(Math.floor(textX), Math.floor(textY));
			this._text.set_size(Math.floor(cloneWidth+100), Math.floor(cloneHeight+100));
			this._text.raise_top();
			});

	workspaceInjections['_init'] = injectToFunction(Workspace.Workspace.prototype, '_init', function(metaWorkspace) {
			if (metaWorkspace && metaWorkspace.index() < 9) {
			createdActors.push(this._tip = new St.Label({ style_class: 'extension-windowsTab-window-tooltip',
					visible: false }));

			this.actor.add_actor(this._tip);
			let signalId = this.actor.connect('notify::scale-x', Lang.bind(this, function() {
					this._tip.set_scale(1 / this.actor.scale_x, 1 / this.actor.scale_x);
					}));
			connectedSignals.push({ obj: this.actor, id: signalId });
			} else
			this._tip = null;
			});

	workspaceInjections['positionWindows'] = injectToFunction(Workspace.Workspace.prototype, 'positionWindows', function(flags) {
			let visibleClones = this._windows.slice();
			if (this._reservedSlot)
			visibleClones.push(this._reservedSlot);

			let slots = this._computeAllWindowSlots(visibleClones.length);
			visibleClones = this._orderWindowsByMotionAndStartup(visibleClones, slots);
			for (let i = 0; i < visibleClones.length; i++) {
			let clone = visibleClones[i];
			let metaWindow = clone.metaWindow;
			let mainIndex = this._lookupIndex(metaWindow);
			let overlay = this._windowOverlays[mainIndex];
			if (overlay)
			overlay.setId(i+1);
			}
			});

	workViewInjections['_init'] = injectToFunction(WorkspacesView.WorkspacesView.prototype, '_init', function(width, height, x, y, workspaces) {
			this._pickWorkspace = false;
			this._pickWindow = false;
			this._keyPressEventId = global.stage.connect('key-press-event', Lang.bind(this, this._onKeyPress));
			this._keyReleaseEventId = global.stage.connect('key-release-event', Lang.bind(this, this._onKeyRelease));
			connectedSignals.push({ obj: global.stage, id: this._keyPressEventId });
			connectedSignals.push({ obj: global.stage, id: this._keyReleaseEventId });
			});

	workViewInjections['_onDestroy'] = injectToFunction(WorkspacesView.WorkspacesView.prototype, '_onDestroy', function() {
			global.stage.disconnect(this._keyPressEventId);
			global.stage.disconnect(this._keyReleaseEventId);
			connectedSignals = [ ];
			});
}

function removeInjection(object, injection, name) {
	if (injection[name] === undefined)
		delete object[name];
	else
		object[name] = injection[name];
}

function disable() {
	for (i in workspaceInjections)
		removeInjection(Workspace.Workspace.prototype, workspaceInjections, i);
	for (i in winInjections)
		removeInjection(Workspace.WindowOverlay.prototype, winInjections, i);
	for (i in workViewInjections)
		removeInjection(WorkspacesView.WorkspacesView.prototype, workViewInjections, i);

	for each (i in connectedSignals)
		i.obj.disconnect(i.id);

	for each (i in createdActors)
		i.destroy();

	resetState();
}

function init() {
	/* do nothing */
}




src/tweak_extension/status-area-horizontal-spacing@mathematical.coffee.gmail.com/prefs.js
const GLib = imports.gi.GLib;
const GObject = imports.gi.GObject;
const Gio = imports.gi.Gio;
const Gtk = imports.gi.Gtk;
const Lang = imports.lang;

const Gettext = imports.gettext.domain('gnome-shell-extensions');
const _ = Gettext.gettext;

const ExtensionUtils = imports.misc.extensionUtils;
const Me = ExtensionUtils.getCurrentExtension();
const Convenience = Me.imports.convenience;

function init() {
    Convenience.initTranslations("gnome-shell-extensions");
}

const StatusAreaHorizontalSpacingPrefsWidget = new GObject.Class({
    Name: 'StatusAreaHorizontalSpacing.Prefs.Widget',
    GTypeName: 'StatusAreaHorizontalSpacingPrefsWidget',
    Extends: Gtk.Grid,

    _init: function(params) {
        this.parent(params);
            this.margin = this.row_spacing = this.column_spacing = 10;

        this._settings = Convenience.getSettings();
        this.attach(new Gtk.Label({ label: _("Horizontal Padding") }), 0, 0, 1, 1);
        let hscale = Gtk.Scale.new_with_range(Gtk.Orientation.HORIZONTAL, 0, 12, 1);
            hscale.set_value(this._settings.get_int('hpadding'));
            hscale.set_digits(0);
            hscale.set_hexpand(true);
            hscale.connect('value-changed', Lang.bind(this, this._onHpaddingChanged));
        this.attach(hscale, 1, 0, 1, 1);
        this._hscale = hscale;
    },

    _onHpaddingChanged: function (hscale) {
        this._settings.set_int('hpadding', this._hscale.get_value());
        //log('hpadding changed! ' + this._settings.get_int('hpadding'));
    }
});

function buildPrefsWidget() {
    let widget = new StatusAreaHorizontalSpacingPrefsWidget();
    widget.show_all();

    return widget;
}


installer.sh
#!/bin/sh

sudo add-apt-repository universe
sudo add-apt-repository ppa:danjaredg/jayatana
sudo apt update

sudo apt install git gnome-tweak-tool gir1.2-clutter-1.0 gir1.2-clutter-gst-3.0 gir1.2-gtkclutter-1.0 jayatana -y
sudo apt install git -y

#Dependency needed for global menu
sudo apt install unity-gtk2-module unity-gtk3-module -y
sudo apt install x11-utils -y

#Needed to recompile gschema of gnome extension with custom default settings
sudo apt-get install libglib2.0-dev -y


git clone https://github.com/RabbitHols/Power-Gnome $HOME/Power-Gnome


sudo cp -r $HOME/Power-Gnome/src/style/themes/* /usr/share/themes
sudo cp -r $HOME/Power-Gnome/src/style/icons/* /usr/share/icons
sudo cp -r $HOME/Power-Gnome/src/style/cursor/* /usr/share/icons

sudo cp -r $HOME/Power-Gnome/src/style/themes/* /usr/share/themes
sudo cp -r $HOME/Power-Gnome/src/style/icons/* /usr/share/icons
sudo cp -r $HOME/Power-Gnome/src/style/cursor/* /usr/share/icons


mkdir $HOME/.local/share/gnome-shell/extensions

sudo yes | cp -rf $HOME/Power-Gnome/src/tweak_extension/user-theme@gnome-shell-extensions.gcampax.github.com $HOME/.local/share/gnome-shell/extensions
sudo yes | cp -rf $HOME/Power-Gnome/src/tweak_extension/activities-config@nls1729 $HOME/.local/share/gnome-shell/extensions

sudo mkdir -p ~/.local/share/gnome-shell/extensions/

#Global Menu
git clone https://gitlab.com/lestcape/Gnome-Global-AppMenu
sudo yes | cp -rf Gnome-Global-AppMenu/gnomeGlobalAppMenu@lestcape ~/.local/share/gnome-shell/extensions/
sudo rm -rf Gnome-Global-AppMenu

sudo rm ~/.local/share/gnome-shell/extensions/gnomeGlobalAppMenu@lestcape/settings-schema.json
sudo cp $HOME/Power-Gnome/src/tweak_extension/gnomeGlobalAppMenu@lestcape/settings-schema.json ~/.local/share/gnome-shell/extensions/gnomeGlobalAppMenu@lestcape/


#Unite
sudo yes | cp -rf $HOME/Power-Gnome/src/tweak_extension/unite@hardpixel.eu ~/.local/share/gnome-shell/extensions/

#Update indicator
sudo yes | cp -rf $HOME/Power-Gnome/src/tweak_extension/apt-update-indicator@franglais125.gmail.com ~/.local/share/gnome-shell/extensions/

#Status area horizontal
sudo yes | cp -rf $HOME/Power-Gnome/src/tweak_extension/status-area-horizontal-spacing@mathematical.coffee.gmail.com ~/.local/share/gnome-shell/extensions/

#App Folder
sudo yes | cp -rf $HOME/Power-Gnome/src/tweak_extension/appfolders-manager@maestroschan.fr ~/.local/share/gnome-shell/extensions/

#App Folder
sudo yes | cp -rf $HOME/Power-Gnome/src/tweak_extension/openweather-extension@jenslody.de ~/.local/share/gnome-shell/extensions/



sudo cp $HOME/Power-Gnome/src/scripts/* /usr/bin
sudo chmod +x /usr/bin/extension-enabler



extension-enabler -e user-theme@gnome-shell-extensions.gcampax.github.com
extension-enabler -e activities-config@nls1729
extension-enabler -e unite@hardpixel.eu
extension-enabler -e gnomeGlobalAppMenu@lestcape
extension-enabler -e apt-update-indicator@franglais125.gmail.com
extension-enabler -e status-area-horizontal-spacing@mathematical.coffee.gmail.com
extension-enabler -e appfolders-manager@maestroschan.fr

sudo cp $HOME/.local/share/gnome-shell/extensions/user-theme@gnome-shell-extensions.gcampax.github.com/schemas/org.gnome.shell.extensions.user-theme.gschema.xml /usr/share/glib-2.0/schemas && sudo glib-compile-schemas /usr/share/glib-2.0/schemas

gsettings set org.gnome.desktop.interface gtk-theme "PowerGnomeGtk" 
gsettings set org.gnome.shell.extensions.user-theme name "PowerGnomeShell"
gsettings set org.gnome.desktop.interface icon-theme 'power-gnome-icons'
gsettings set org.gnome.desktop.interface cursor-theme "power-gnome-cursor"

dconf reset -f /org/gnome/shell/extensions/dash-to-dock/
gsettings set org.gnome.shell.extensions.dash-to-dock show-apps-at-top true
gsettings set org.gnome.shell.extensions.dash-to-dock custom-theme-shrink false

gsettings set org.gnome.desktop.wm.preferences button-layout 'close,maximize,minimize:'
gsettings set org.gnome.shell.extensions.dash-to-dock dash-max-icon-size 30


# For now no custom wallpaper
#sudo mv $HOME/Power-Gnome/src/style/wallpaper/powerGnome.png /usr/share/backgrounds/warty-final-ubuntu.png
#gsettings set org.gnome.desktop.background picture-uri file:////usr/share/backgrounds/warty-final-ubuntu.png

#Adjust schema error

sudo cp $HOME/.local/share/gnome-shell/extensions/activities-config@nls1729/schemas/org.gnome.shell.extensions.activities-config.gschema.xml /usr/share/glib-2.0/schemas/ 
sudo glib-compile-schemas /usr/share/glib-2.0/schemas/

sudo cp $HOME/.local/share/gnome-shell/extensions/unite@hardpixel.eu/schemas/org.gnome.shell.extensions.unite.gschema.xml /usr/share/glib-2.0/schemas/ 
sudo glib-compile-schemas /usr/share/glib-2.0/schemas/

sudo rm -r Power-Gnome

dbus-send --type=method_call --print-reply --dest=org.gnome.Shell /org/gnome/Shell org.gnome.Shell.Eval string:'global.reexec_self()'
sleep 5s && dbus-send --type=method_call --print-reply --dest=org.gnome.Shell /org/gnome/Shell org.gnome.Shell.Eval string:'global.reexec_self()'
clear
gnome-session-quit
echo Done !!
